Adding missing documentation and examples (#118)
rsksmart/rif-relay: docs/integration_guide.md

        tokenContract,
        tokenAmount,
        data: "0x",
        index: smartWalletIndex,
        recoverer: ZERO_ADDRESS,
        isSmartWalletDeploy: true,
        onlyPreferredRelays: true,
        smartWalletAddress
    });
    
    // relay transaction
    const unsigned_tx = {
        // some common web3 transaction with the common parameters.
    };

    const tokenAmountForRelay = "10";
    
    const relayTransaction = web3.eth.sendTransaction({
        from: account.address,
        callVerifier: relayVerifierAddress,
        callForwarder: smartWalletAddress,
        isSmartWalletDeploy: false,
        onlyPreferredRelays: true,
        tokenAmount: tokenAmountForRelay,
        tokenContract,
        ...unsigned_tx,
    });
```

**Note: in the example above the `account` object is assumed as an object containing the address (as string) and 
the privateKey (as buffer)**

Before running this example, you need to know a few requirements:

1. The smart wallet address generated by the contract call should be funded with tokens before running the deploy call or
   you can set tokenAmount to 0 (or remove it) to make a subsidized deploy instead.

> 2. The token address you use should be allowed, to allow it you need to make a call to the contracts involved to allow
>    working with your token. These contracts are the relay and deploy verifiers, and the method is `acceptToken`, it should be called

Reviewer: ```suggestion
2. The token address you use need to be explicitly allowed. To do so, make a call to the contracts involved to allow them to work with your particular token. These contracts are the relay and deploy verifiers, and the method is `acceptToken`, it should be called
```

Another option is to use Enveloping through a Relay Provider. The latter wraps web3, and then all transactions and calls are made through the Relay Provider. If a Relay Client is not provided then the Relay Provider creates an instance.

```typescript
    import { RelayProvider, resolveConfiguration } from "@rsksmart/enveloping";
    import Web3 from "web3";

    const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";

    const web3 = new Web3("http://localhost:4444");
    
    const smartWalletFactoryAbi = {};// some json containing the abi of the smart wallet factory contract.

>     const smartWalletFactoryAddress = "0x3bA95e1cccd397b5124BcdCC5bf0952114E6A701"; // the smart wallet factort contract address (can be retrieved from the summary of the deployment).

Reviewer: (valid for the rest of the doc)

>     const smartWalletFactoryAddress = "0x3bA95e1cccd397b5124BcdCC5bf0952114E6A701"; // the smart wallet factort contract address (can be retrieved from the summary of the deployment).

Author: Is not, but it will be, and also the actual output of the deployment shows you the addresses of the contracts so is not to far of a summary. Do you still want to change that comment?

>     const smartWalletFactoryAddress = "0x3bA95e1cccd397b5124BcdCC5bf0952114E6A701"; // the smart wallet factort contract address (can be retrieved from the summary of the deployment).

Reviewer: nope, with what you're saying in mind, i think it's fine as it is :+1:

        tokenContract,
        tokenAmount,
        data: "0x",
        index: smartWalletIndex,
        recoverer: ZERO_ADDRESS,
        isSmartWalletDeploy: true,
        onlyPreferredRelays: true,
        smartWalletAddress
    });
    
    // relay transaction
    const unsigned_tx = {
        // some common web3 transaction with the common parameters.
    };

    const tokenAmountForRelay = "10";
    
    const relayTransaction = web3.eth.sendTransaction({
        from: accountAddress,
        callVerifier: relayVerifierAddress,
        callForwarder: smartWalletAddress,
        isSmartWalletDeploy: false,
        onlyPreferredRelays: true,
        tokenAmount: tokenAmountForRelay,
        tokenContract,
        ...unsigned_tx,
    });
```

Before running this example, you need to know a few requirements:

1. The smart wallet address generated by the contract call should be funded with tokens before running the deploy call.
2. The token address you use should be allowed, to do that you need to make a call to the contracts involved to allow
   working with your token, only the owner of the contracts can do that, but if you are running this on regtest, then the account[0]

>    is the owner and you can allow tokens by calling the relay verifier and deploy verifier (for both wallets, smart wallet and custom smart wallet) contracts manually with web3.

Reviewer: please break up this sentence into multiple ones, it's a bit hard to read as it is

Reviewer: also, account[0] should be `accounts[0]`

Reviewer: what if you are not running this on regtest? we should guide the user in this case too

Author: Should be the same for all chains, the only thing that changes is that in testnet and mainnet instead of you being allowing the tokens you have to use one of the already allowed tokens.

        const smartWalletDeployVerifier = await new web3.eth.Contract(smartWalletDeployVerifierAbi, deployVerifierAddress);
        const smartWalletRelayVerifier = await new web3.eth.Contract(relayVerifierAbi, relayVerifierAddress);
        const customSmartWalletDeployVerifier = await new web3.eth.Contract(customSmartWalletDeployVerifierAbi, customDeployVerifierAddress);
        const customSmartWalletRelayVerifier = await new web3.eth.Contract(relayVerifierAbi, customRelayVerifierAddress);
        const accounts = await web3.eth.getAccounts();
        
        const tokenAddress = "0x0E569743F573323F430B6E14E5676EB0cCAd03D9"; // token address to allow
    
        await smartWalletDeployVerifier.methods.acceptToken(tokenAddress).send({from: accounts[0]});
        await smartWalletRelayVerifier.methods.acceptToken(tokenAddress).send({from: accounts[0]});
        await customSmartWalletDeployVerifier.methods.acceptToken(tokenAddress).send({from: accounts[0]});
        await customSmartWalletRelayVerifier.methods.acceptToken(tokenAddress).send({from: accounts[0]});
   ```

> 3. Some times you could have troubles with gas estimation on relaying transactions. One thing you can do to

Reviewer: ```suggestion
3. Sometimes you can have problems with gas estimation when relaying transactions. One thing you can do to
```

> avoid those problems is to increase a parameter that enveloping uses as cushion for the gas spend by the relay transactions.

Reviewer: ```suggestion
avoid those problems is to increase a parameter that enveloping uses as cushion for the gas spent by the relay transactions.
```

Another option is to use Enveloping through a Relay Provider. The latter wraps web3, and then all transactions and calls are made through the Relay Provider. If a Relay Client is not provided then the Relay Provider creates an instance.

```typescript
    import { RelayProvider, resolveConfiguration } from "@rsksmart/enveloping";
    import Web3 from "web3";

    const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";

    const web3 = new Web3("http://localhost:4444");
    
    const smartWalletFactoryAbi = {};// some json containing the abi of the smart wallet factory contract.
    const smartWalletFactoryAddress = "0x3bA95e1cccd397b5124BcdCC5bf0952114E6A701"; // the smart wallet factort contract address (can be retrieved from the summary of the deployment).
    const smartWalletIndex = 0; // the index of the smart wallet
    const accountAddress = "0xB6009fD7a09cB4C6846ab19Be81d7D0b533cF8B7"; // your account address

>     const accountPrivateKey = "0xc2decf57baed5c1f4e546dc04eab97edb753c72723720af95f14c3ac1e3ef830"; // your account private key

Author: Yes we know but to do that outside enveloping you need to have the AccountManager exported and it's not so you can't import it to instantiate it, so based on that we can't do it because is impossible to get the definition from the enveloping library, that's something we should improve (exporting resources) anyway, but we know is not a good idea this was just a working example.

>     const accountPrivateKey = "0xc2decf57baed5c1f4e546dc04eab97edb753c72723720af95f14c3ac1e3ef830"; // your account private key

Reviewer 2: I suggest to fix the export problem, or at least to show a snippet of code with the potential solution instead of suggesting the usage of the private key in this way. If you take a look at the demo we created for the webinar, it is possible to import the different Enveloping components from the dist folder once it is built. The only missing part is the creation and publishing of an npm  package to make it easier to be imported.

>     const accountPrivateKey = "0xc2decf57baed5c1f4e546dc04eab97edb753c72723720af95f14c3ac1e3ef830"; // your account private key

Author: Make sense, i will add an external reference and put a note to let the user know that it should be externally managed instead of putting the privateKey directly

    // deploy smart wallet
    const deployTransaction = await provider.deploySmartWallet({
        from: accountAddress,
        to: ZERO_ADDRESS,
        gas: "0x27100",
        value: "0",
        callVerifier: deployVerifierAddress,
        callForwarder: smartWalletFactoryAddress,
        tokenContract,
        tokenAmount,
        data: "0x",
        index: smartWalletIndex,
        recoverer: ZERO_ADDRESS,
        isSmartWalletDeploy: true,
        onlyPreferredRelays: true,
        smartWalletAddress
    });
    
    // relay transaction
    const unsigned_tx = {
        // some common web3 transaction with the common parameters.
    };

    const tokenAmountForRelay = "10";
    
    const relayTransaction = web3.eth.sendTransaction({
        from: accountAddress,
        callVerifier: relayVerifierAddress,
        callForwarder: smartWalletAddress,
        isSmartWalletDeploy: false,
        onlyPreferredRelays: true,
        tokenAmount: tokenAmountForRelay,
        tokenContract,
        ...unsigned_tx,
    });
```

Before running this example, you need to know a few requirements:

> 1. The smart wallet address generated by the contract call should be funded with tokens before running the deploy call.

Author: We tried this but the contract was complaining about it and we had to fund the smart wallet address before deploying it due to the requirements on it.

> 1. The smart wallet address generated by the contract call should be funded with tokens before running the deploy call.

Author: This is correct, i will add as an alternative that the tokenAmount can be 0 to make a subsidized deploy.
